sdk/core/azure-core/CMakeLists.txt

```
@@ -7,6 +7,7 @@ project(azure-core LANGUAGES CXX)
 set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CXX_STANDARD_REQUIRED True)
 set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
+add_compile_definitions(AZ_CORE_EXPORTS)
```

Since this is only relevant for msvc and windows, can we add a windows suffix both here and to the export.hpp file name. Also, export is somewhat generic and could be overloaded or cause confusion:
```suggestion
add_compile_definitions(AZ_CORE_DLL_EXPORTS_WINDOWS)
```

---

I believe things like this will cause issues on Windows sometimes. We only need the dllexport/dllimport for dynamic libraries, not static libraries.

---

You can take reference from what we're doing for cpplite

https://github.com/Azure/azure-storage-cpplite/blob/master/include/storage_EXPORTS.h
https://github.com/Azure/azure-storage-cpplite/blob/be490edaf413dc113d8182cbc7a29140d3a63481/CMakeLists.txt#L157

---

I have updated our approach, it is better. You use `defined(_WINDLL)`, and it does not work when customer is building a DLL or exe, but is linking with a static azure SDK library.

The way I put it now is that `declspec(dllimport/export)` only exists in Windows builds when the library is being built as DLL, and is a no-op in all other cases.
Then,
* we either carry the knowledge that the SDK lib was built as Windows DLL via `target_compile_definitions(lib PUBLIC macro)` - this is the case when a customer consumes our library via `fetchcontent()`
* or, when we install the library as a VcPackage, install step would patch the `dll_import_export.hpp` header when the library is being built as Windows DLL. And there is no conflict there, because on VcPkg, if you build for both static and DLL, your headers are in two separate locations of their own.